src/dns.rs

```diff
@@ -0,0 +1,510 @@
+use crate::runtime::Runtime;
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::error;
+use std::fmt;
+use std::io;
+use std::net::AddrParseError;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver {
+    resolver: Option<AsyncResolver>,
+}
+
+impl DnsResolver {
+    pub fn new<'a, R: Into<Option<&'a mut Runtime>>>(
+        config: &GlobalOptions,
+        runtime: R,
+    ) -> Result<Self, DnsError> {
+        if config.dns.is_empty() {
+            return Ok(DnsResolver::default());
+        }
+
+        let mut resolve_config = ResolverConfig::new();
+
+        let mut errors = Vec::new();
+
+        for s in config.dns.iter() {
+            let parsed = s
+                .parse::<SocketAddr>()
+                .or_else(|_| s.parse::<IpAddr>().map(|ip| SocketAddr::new(ip, DNS_PORT)));
+            match parsed {
+                Ok(socket_addr) => resolve_config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol: Protocol::Udp,
+                    tls_dns_name: None,
+                }),
+                Err(error) => errors.push((s.clone(), error)),
+            }
+        }
+
+        if !errors.is_empty() {
+            return Err(DnsError::InputError { servers: errors });
+        }
+
+        let mut options = ResolverOpts::default();
+        options.attempts = 2;
+        options.validate = false;
+        options.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+        options.num_concurrent_reqs = 1;
+
+        let (resolver, worker) = AsyncResolver::new(resolve_config, options);
+
+        if let Some(runtime) = runtime.into() {
+            runtime.spawn(worker);
+        } else {
+            // spin up worker thread for worker future
+            std::thread::spawn(move || {
+                // Worker futer will finish when all DnsResolvers are dropped.
+                // Then this thread will also end.
+                tokio::runtime::current_thread::run(worker);
+            });
+        }
+
+        Ok(DnsResolver {
+            resolver: Some(resolver),
+        })
+    }
+
+    /// Resolves host:port address
+    /// Should be only called from outside tokio. From inside tokio, Async version should be used.
+    pub fn resolve_address(&self, s: &str) -> Result<IntoIter<SocketAddr>, DnsError> {
```

Comment:
I think we should just return a future to keep this module simple and then provide a `crate::runtime::block_on` that uses https://docs.rs/tokio/0.1.22/tokio/runtime/current_thread/struct.Runtime.html#method.block_on so that it is trivial for a non tokio context to call this. What do you think?

Reply:
That's a better way.